wincode-derive/src/schema_read.rs

```diff
@@ -225,27 +230,249 @@ fn impl_struct_extensions(args: &SchemaArgs) -> Result<TokenStream> {
+            }
+        });
+        quote! {
+            impl #builder_impl_generics Drop for #builder_ident #builder_ty_generics #builder_where_clause {
+                fn drop(&mut self) {
+                    let dst_ptr = self.inner.as_mut_ptr();
+                    #(#drops)*
+                }
+            }
         }
-    ))
+    };
+
+    let builder_impl = {
+        let is_fully_init_mask = if fields.len() as u32 == builder_bit_set_bits {
+            quote!(#builder_bit_set_ty::MAX)
+        } else {
+            let field_bits = LitInt::new(&fields.len().to_string(), Span::call_site());
+            quote!(((1 as #builder_bit_set_ty) << #field_bits) - 1)
+        };
+
+        quote! {
+            impl #builder_impl_generics #builder_ident #builder_ty_generics #builder_where_clause {
+                #vis const fn from_maybe_uninit_mut(inner: &'_wincode_inner mut MaybeUninit<#builder_dst>) -> Self {
+                    Self {
+                        inner,
+                        init_set: 0,
+                    }
+                }
+
+                /// Check if the builder is fully initialized.
+                ///
+                /// This will check if all field initialization bits are set.
+                #[inline]
+                #vis const fn is_init(&self) -> bool {
+                    self.init_set == #is_fully_init_mask
+                }
+
+                /// Assume the builder is fully initialized, and return a mutable reference to the inner `MaybeUninit` struct.
+                ///
+                /// The builder will be forgotten, so the drop logic will not longer run.
+                ///
+                /// # Safety
+                ///
+                /// Calling this when the content is not yet fully initialized causes undefined behavior: it is up to the caller
+                /// to guarantee that the `MaybeUninit<T>` really is in an initialized state.
+                #[inline]
+                #vis const unsafe fn into_assume_init_mut(mut self) -> &'_wincode_inner mut #builder_dst {
+                    // SAFETY: reference lives beyond the scope of the builder, and builder is forgotten.
+                    let inner = unsafe { ptr::read(&mut self.inner) };
+                    mem::forget(self);
+                    // SAFETY: Caller asserts the `MaybeUninit<T>` is in an initialized state.
+                    unsafe {
+                        inner.assume_init_mut()
+                    }
+                }
+
+                /// Forget the builder, disabling the drop logic.
+                ///
+                /// # Safety
+                ///
+                /// Assuming the source `MaybeUninit<T>` when the content is not yet fully initialized causes undefined behavior.
+                /// It is up to the caller to guarantee that the `MaybeUninit<T>` really is in an initialized state.
+                #[inline]
+                #vis const unsafe fn assume_init_forget(self) {
```

Comment:
Not sure I like this name either... since it doesn't actually run the `assume_init*` on underlying `MaybeUninit`, which the name might suggest.
Maybe it should be something like `finish_init_forget` or `init_done_forget` ?

We could make it return the underlying `&mut MaybeUnnit` and call it `into_inner_done`, which might be convenient in some cases, but also be a bit confusing.

Reply:
`finish` seems fine, not sure it really needs `unsafe`, since it doesn't do anything unsafe by itself and the user should be aware of need to initialize all parts of memory when starting to use the builder in the first place. It's also a call that always needs to be made unless `into_assume_init_mut` is used, so forcing to add another `unsafe` block seems off.

 I would be in fact more concerned with how to ensure user do not forget to call `finished`.